File: LeetCode (en comments)/54. Spiral Matrix.py
class Solution:
    def spiralOrder(self, arr: List[List[int]]) -> List[int]:
        scr = []
        m = len(arr[0])
        n = len(arr)
        top, bottom, left, right = 0, n - 1, 0, m - 1

        while len(scr) < n*m:

            for i in range(left, right + 1):
                scr.append(arr[top][i])
            top += 1

            for j in range(top, bottom + 1):
                scr.append(arr[j][right])
            right -= 1

            if top <= bottom:
                for i in range(right, left - 1, -1):
                    scr.append(arr[bottom][i])
                bottom -= 1

            if left <= right:
                for j in range(bottom, top - 1, -1):
                    scr.append(arr[j][left])
                left += 1

        return scr

File: LeetCode (en comments)/54. Spiral Matrix.ts
function spiralOrder(matrix: number[][]): number[] {
    const result: number[] = [];
    let top: number = 0;
    let bottom: number = matrix.length - 1;
    let left: number = 0;
    let right: number = matrix[0].length - 1;

    while (result.length < matrix.length * matrix[0].length) {
        for (let i = left; i <= right; i++) {
            result.push(matrix[top][i]);
        }
        top++;

        for (let i = top; i <= bottom; i++) {
            result.push(matrix[i][right]);
        }
        right--;

        if (top <= bottom) {
            for (let i = right; i >= left; i--) {
                result.push(matrix[bottom][i]);
            }
            bottom--;
        }

        if (left <= right) {
            for (let i = bottom; i >= top; i--) {
                result.push(matrix[i][left]);
            }
            left++;
        }
    }

    return result;
}